File: scripts/llm-hooks/hooklib.py
#!/usr/bin/env python3
"""Small helpers shared by Codex and Claude Code command hooks."""

from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path
from typing import Any


def read_event() -> dict[str, Any]:
    raw = sys.stdin.read().strip()
    if not raw:
        return {}
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return {"raw_stdin": raw}
    return value if isinstance(value, dict) else {"value": value}


def repo_root(event: dict[str, Any]) -> Path:
    cwd = event.get("cwd") or os.environ.get("CLAUDE_PROJECT_DIR") or os.getcwd()
    path = Path(str(cwd)).resolve()
    for candidate in [path, *path.parents]:
        if (candidate / ".git").exists():
            return candidate
    return path


def tool_name(event: dict[str, Any]) -> str:
    return str(event.get("tool_name") or event.get("toolName") or "")


def tool_input(event: dict[str, Any]) -> dict[str, Any]:
    value = event.get("tool_input") or event.get("toolInput") or {}
    return value if isinstance(value, dict) else {}


def command_text(event: dict[str, Any]) -> str:
    data = tool_input(event)
    candidates = [
        data.get("command"),
        data.get("cmd"),
        data.get("script"),
        data.get("url"),
        data.get("text"),
        data.get("element"),
        data.get("target"),
        data.get("value"),
        event.get("command"),
    ]
    return "\n".join(str(item) for item in candidates if item)


def edited_paths(event: dict[str, Any]) -> list[str]:
    data = tool_input(event)
    keys = ("file_path", "path", "filename", "target_file", "notebook_path")
    paths = [str(data[key]) for key in keys if data.get(key)]

    raw = json.dumps(data, ensure_ascii=False)
    for match in re.finditer(r"(data/companies/[^\"'\s]+|data/tracker\.md|candidate/cv/[^\"'\s]+)", raw):
        paths.append(match.group(1))

    seen: set[str] = set()
    result: list[str] = []
    for path in paths:
        if path not in seen:
            seen.add(path)
            result.append(path)
    return result


def print_claude_deny(reason: str, event_name: str = "PreToolUse") -> None:
    payload = {
        "hookSpecificOutput": {
            "hookEventName": event_name,
            "permissionDecision": "deny",
            "permissionDecisionReason": reason,
        }
    }
    print(json.dumps(payload))


def warn(reason: str) -> None:
    print(reason, file=sys.stderr)

File: scripts/llm-hooks/hooklib.js
import { existsSync, readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';

export function readEvent() {
  const raw = readFileSync(0, 'utf8').trim();
  if (!raw) return {};
  try {
    const value = JSON.parse(raw);
    return value && typeof value === 'object' && !Array.isArray(value) ? value : { value };
  } catch {
    return { raw_stdin: raw };
  }
}

export function repoRoot(event = {}) {
  const cwd = event.cwd || process.env.CLAUDE_PROJECT_DIR || process.cwd();
  let path = resolve(String(cwd));
  while (true) {
    if (existsSync(resolve(path, '.git'))) return path;
    const parent = dirname(path);
    if (parent === path) return resolve(String(cwd));
    path = parent;
  }
}

export function toolName(event) {
  return String(event.tool_name || event.toolName || '');
}

export function toolInput(event) {
  const value = event.tool_input || event.toolInput || {};
  return value && typeof value === 'object' && !Array.isArray(value) ? value : {};
}

export function commandText(event) {
  const data = toolInput(event);
  const candidates = [
    data.command,
    data.cmd,
    data.script,
    data.url,
    data.text,
    data.element,
    data.target,
    data.value,
    event.command,
  ];
  return candidates.filter(Boolean).map(String).join('\n');
}

export function editedPaths(event) {
  const data = toolInput(event);
  const keys = ['file_path', 'path', 'filename', 'target_file', 'notebook_path'];
  const paths = keys.filter((key) => data[key]).map((key) => String(data[key]));

  const raw = JSON.stringify(data);
  for (const match of raw.matchAll(/(data\/companies\/[^"'\s]+|data\/tracker\.md|candidate\/cv\/[^"'\s]+|candidate\/application-answers\.md)/g)) {
    paths.push(match[1]);
  }

  return [...new Set(paths)];
}

export function printClaudeDeny(reason, eventName = 'PreToolUse') {
  console.log(JSON.stringify({
    hookSpecificOutput: {
      hookEventName: eventName,
      permissionDecision: 'deny',
      permissionDecisionReason: reason,
    },
  }));
}

export function warn(reason) {
  console.error(reason);
}
